File: platforms/android/module/src/main/kotlin/rs/whisker/runtime/WhiskerModuleRegistrar.kt
package rs.whisker.runtime

import android.content.Context
import android.util.Log
import android.view.View
import java.util.concurrent.ConcurrentHashMap

public typealias WhiskerElementEventSink = (WhiskerEventBinding, WhiskerValue) -> Unit

/** Taffy available-space constraint on one measurement axis. */
public enum class WhiskerAvailableSpace { DEFINITE, MIN_CONTENT, MAX_CONTENT }

public data class WhiskerMeasureRequest(
    val availableWidth: Float?, val availableHeight: Float?,
    val availableWidthKind: WhiskerAvailableSpace,
    val availableHeightKind: WhiskerAvailableSpace,
    val knownWidth: Float?, val knownHeight: Float?,
    val payloadVersion: Int, val payload: WhiskerValue,
)

public data class WhiskerMeasuredSize(val width: Float, val height: Float)

/** Decoded `SetText` payload passed to the Host element implementation. */
public data class WhiskerTextContent(
    public val value: String,
    public val fontSize: Float,
    public val fontWeight: Int,
    public val fontFamilies: List<String> = listOf("system"),
    public val fontStyle: WhiskerFontStyle = WhiskerFontStyle.NORMAL,
    public val lineHeight: Float? = null,
    public val letterSpacing: Float = 0f,
    public val fontFeatures: List<WhiskerFontFeature> = emptyList(),
    public val fontVariations: List<WhiskerFontVariation> = emptyList(),
    public val fontOpticalSizing: WhiskerFontOpticalSizing = WhiskerFontOpticalSizing.NONE,
    public val color: Int,
    public val direction: WhiskerTextDirection = WhiskerTextDirection.AUTO,
    public val alignment: WhiskerTextAlignment = WhiskerTextAlignment.START,
    public val indent: WhiskerTextIndent = WhiskerTextIndent(),
    public val wrap: Boolean = true,
    public val wordBreak: WhiskerTextWordBreak = WhiskerTextWordBreak.NORMAL,
    public val maxLines: Int = 0,
    public val overflow: WhiskerTextOverflow = WhiskerTextOverflow.CLIP,
    public val decoration: WhiskerTextDecoration? = null,
    public val shadow: WhiskerTextShadow? = null,
)

/** Resolved inherited text style delivered independently from text content. */
public data class WhiskerTextStyle(
    public val fontSize: Float,
    public val fontWeight: Int,
    public val fontFamilies: List<String> = listOf("system"),
    public val fontStyle: WhiskerFontStyle = WhiskerFontStyle.NORMAL,
    public val lineHeight: Float? = null,
    public val letterSpacing: Float = 0f,
    public val fontFeatures: List<WhiskerFontFeature> = emptyList(),
    public val fontVariations: List<WhiskerFontVariation> = emptyList(),
    public val fontOpticalSizing: WhiskerFontOpticalSizing = WhiskerFontOpticalSizing.NONE,
    public val color: Int,
    public val direction: WhiskerTextDirection = WhiskerTextDirection.AUTO,
    public val alignment: WhiskerTextAlignment = WhiskerTextAlignment.START,
    public val decoration: WhiskerTextDecoration? = null,
    public val shadow: WhiskerTextShadow? = null,
)

public fun WhiskerTextContent.styleSnapshot(): WhiskerTextStyle = WhiskerTextStyle(
    fontSize = fontSize,
    fontWeight = fontWeight,
    fontFamilies = fontFamilies,
    fontStyle = fontStyle,
    lineHeight = lineHeight,
    letterSpacing = letterSpacing,
    fontFeatures = fontFeatures,
    fontVariations = fontVariations,
    fontOpticalSizing = fontOpticalSizing,
    color = color,
    direction = direction,
    alignment = alignment,
    decoration = decoration,
    shadow = shadow,
)

public data class WhiskerFontFeature(public val tag: String, public val value: Long)
public data class WhiskerFontVariation(public val tag: String, public val value: Float)
public enum class WhiskerFontOpticalSizing { AUTO, NONE }
public enum class WhiskerFontStyle { NORMAL, ITALIC, OBLIQUE }

public enum class WhiskerTextDirection { AUTO, LEFT_TO_RIGHT, RIGHT_TO_LEFT }

public enum class WhiskerTextAlignment { START, END, LEFT, RIGHT, CENTER }

public enum class WhiskerTextWordBreak { NORMAL, BREAK_ALL, KEEP_ALL }

public enum class WhiskerTextOverflow { CLIP, ELLIPSIS }

/** First-line indentation; percentage is relative to the final Text width. */
public data class WhiskerTextIndent(
    public val logicalPixels: Float = 0f,
    public val percentage: Float = 0f,
)

/** One inherited Whisker text decoration. */
public data class WhiskerTextDecoration(
    public val line: WhiskerTextDecorationLine,
    public val style: WhiskerTextDecorationStyle,
    public val color: Int,
)

public enum class WhiskerTextDecorationLine { UNDERLINE, LINE_THROUGH }

public enum class WhiskerTextDecorationStyle { SOLID, DOUBLE, DOTTED, DASHED, WAVY }

/** One resolved shadow painted behind native text glyphs. */
public data class WhiskerTextShadow(
    public val offsetX: Float,
    public val offsetY: Float,
    public val blurRadius: Float,
    public val color: Int,
)

/** One native View mounted for a retained Whisker node. */
public class WhiskerMountedElement internal constructor(
    public val registration: WhiskerElementRegistration,
    public val view: View,
    private val textUpdater: ((View, WhiskerTextContent) -> Unit)?,
    private val textStyleUpdater: ((View, WhiskerTextStyle) -> Unit)?,
    private val childrenHost: ((View) -> android.view.ViewGroup)?,
    private val properties: Map<Int, WhiskerPropComponent>,
    private val commands: Map<Int, WhiskerCommandComponent>,
    private val eventsByName: Map<String, WhiskerEventBinding>,
    eventSink: WhiskerElementEventSink,
) {
    private var eventMask: Long = 0
    private var eventSink: WhiskerElementEventSink = eventSink

    init {
        installEventSink()
    }

    private fun installEventSink() {
        (view as? WhiskerEventSource)?.installWhiskerEventSink { name, detail ->
            val event = eventsByName[name] ?: return@installWhiskerEventSink
            val bit = 1L shl (event.id - 1)
            if (eventMask and bit != 0L) eventSink(event, detail)
        }
    }

    public fun setProperty(id: Int, value: WhiskerValue) {
        properties[id]?.setter?.invoke(view, value)
    }

    /** Clear is a protocol operation and is intentionally not `WhiskerValue.Null`. */
    public fun clearProperty(id: Int) {
        properties[id]?.clearer?.invoke(view)
    }

    public fun invokeCommand(id: Int, parameters: WhiskerValue) {
        commands[id]?.handler?.invoke(view, parameters)
    }

    public fun setEventMask(mask: Long) { eventMask = mask }

    public fun setText(content: WhiskerTextContent): Boolean {
        val update = textUpdater ?: return false
        update(view, content)
        return true
    }

    public fun setTextStyle(style: WhiskerTextStyle): Boolean {
        val update = textStyleUpdater ?: return false
        update(view, style)
        return true
    }

    public fun childrenHost(): android.view.ViewGroup? = childrenHost?.invoke(view)

    /** Resets protocol-owned state before a built-in presentation is reused. */
    public fun prepareForReuse(eventSink: WhiskerElementEventSink) {
        properties.values.forEach { it.clearer(view) }
        eventMask = 0
        this.eventSink = eventSink
        installEventSink()
    }

    public fun dispose() { (view as? WhiskerEventSource)?.installWhiskerEventSink(null) }
}

/** Host-owned declaration. It contains names and behavior, never Rust IDs. */
public class WhiskerElementFactory(
    public val name: String,
    internal val textUpdater: ((View, WhiskerTextContent) -> Unit)? = null,
    internal val textStyleUpdater: ((View, WhiskerTextStyle) -> Unit)? = null,
    internal val childrenHost: ((View) -> android.view.ViewGroup)? = null,
    internal val measurer: ((WhiskerMeasureRequest) -> WhiskerMeasuredSize?)? = null,
    internal val makeView: (Context) -> View,
) {
    init {
        require(name.isNotEmpty() && '@' !in name) {
            "Host element name must be non-empty and versionless"
        }
    }


    internal fun withTextStyleUpdater(
        updater: ((View, WhiskerTextStyle) -> Unit)?,
    ): WhiskerElementFactory = WhiskerElementFactory(
        name = name,
        textUpdater = textUpdater,
        textStyleUpdater = updater ?: textStyleUpdater,
        childrenHost = childrenHost,
        measurer = measurer,
        makeView = makeView,
    )

    internal fun withMeasurer(
        provider: ((WhiskerMeasureRequest) -> WhiskerMeasuredSize?)?,
    ): WhiskerElementFactory = WhiskerElementFactory(
        name = name,
        textUpdater = textUpdater,
        textStyleUpdater = textStyleUpdater,
        childrenHost = childrenHost,
        measurer = provider ?: measurer,
        makeView = makeView,
    )
}

private data class WhiskerDeclaredElement(
    val factory: WhiskerElementFactory,
    val properties: Map<String, WhiskerPropComponent>,
    val events: Set<String>,
    val commands: Map<String, WhiskerCommandComponent>,
)

internal data class WhiskerBoundElement(
    val registration: WhiskerElementRegistration,
    val factory: WhiskerElementFactory,
    val properties: Map<Int, WhiskerPropComponent>,
    val commands: Map<Int, WhiskerCommandComponent>,
)

/** One surface's immutable-after-bootstrap mapping from Rust IDs to Host declarations. */
public class WhiskerElementBindings private constructor() {
    @Volatile private var boundByType: Map<Int, WhiskerBoundElement> = emptyMap()

    internal fun publish(bindings: Map<Int, WhiskerBoundElement>) {
        boundByType = bindings
    }

    public fun mount(
        elementType: Int,
        context: Context,
        eventSink: WhiskerElementEventSink,
    ): WhiskerMountedElement? {
        val element = boundByType[elementType] ?: return null
        val view = element.factory.makeView(context)
        return WhiskerMountedElement(
            element.registration,
            view,
            element.factory.textUpdater,
            element.factory.textStyleUpdater,
            element.factory.childrenHost,
            element.properties,
            element.commands,
            element.registration.events.associateBy { it.name },
            eventSink,
        )
    }

    public fun measure(elementType: Int, request: WhiskerMeasureRequest): WhiskerMeasuredSize? =
        boundByType[elementType]?.factory?.measurer?.invoke(request)

    public fun registration(elementType: Int): WhiskerElementRegistration? =
        boundByType[elementType]?.registration

    internal companion object {
        fun empty(): WhiskerElementBindings = WhiskerElementBindings()
    }
}

/** Process-wide Host declarations used to negotiate independent surface-local binding tables. */
public object WhiskerElementRegistry {
    private const val LOG_TAG = "WhiskerElementRegistry"
    private val declarations = ConcurrentHashMap<String, WhiskerDeclaredElement>()

    @JvmStatic
    public fun newBindings(): WhiskerElementBindings = WhiskerElementBindings.empty()

    @JvmStatic
    public fun register(factory: WhiskerElementFactory) {
        register(factory, emptyMap(), emptySet(), emptyMap())
    }

    internal fun register(view: WhiskerViewComponent, fallbackName: String) {
        val name = view.elementName ?: fallbackName
        view.factory?.let { factory ->
            val properties = view.components.filterIsInstance<WhiskerPropComponent>().associateBy { it.name }
            val events = view.components.filterIsInstance<WhiskerEventsComponent>().flatMap { it.names }.toSet()
            val commands = view.components.filterIsInstance<WhiskerCommandComponent>().associateBy { it.name }
            val textStyle = view.components.filterIsInstance<WhiskerTextStyleComponent>().singleOrNull()
            val measurement = view.components.filterIsInstance<WhiskerMeasurementComponent>().singleOrNull()
            register(
                factory.withTextStyleUpdater(textStyle?.handler).withMeasurer(measurement?.handler),
                properties,
                events,
                commands,
            )
            return
        }
        val declaredClass = requireNotNull(view.viewClass) { "$name View declaration needs a class or factory" }
        require(View::class.java.isAssignableFrom(declaredClass)) {
            "$name View class must extend android.view.View"
        }
        @Suppress("UNCHECKED_CAST")
        val viewClass = declaredClass as Class<out View>
        val properties = view.components.filterIsInstance<WhiskerPropComponent>().associateBy { it.name }
        val events = view.components.filterIsInstance<WhiskerEventsComponent>().flatMap { it.names }.toSet()
        val commands = view.components.filterIsInstance<WhiskerCommandComponent>().associateBy { it.name }
        val textStyle = view.components.filterIsInstance<WhiskerTextStyleComponent>().singleOrNull()
        val measurement = view.components.filterIsInstance<WhiskerMeasurementComponent>().singleOrNull()
        register(
            WhiskerElementFactory(
                name = name,
                textStyleUpdater = textStyle?.handler,
                measurer = measurement?.handler,
                makeView = { context -> viewClass.getConstructor(Context::class.java).newInstance(context) },
            ),
            properties,
            events,
            commands,
        )
    }

    private fun register(
        factory: WhiskerElementFactory,
        properties: Map<String, WhiskerPropComponent>,
        events: Set<String>,
        commands: Map<String, WhiskerCommandComponent>,
    ) {
        require(declarations.putIfAbsent(factory.name, WhiskerDeclaredElement(factory, properties, events, commands)) == null) {
            "element factory already registered for ${factory.name}"
        }
    }

    /** Match Host strings to Rust registrations and compile compact dispatch tables. */
    @JvmStatic
    public fun bind(
        target: WhiskerElementBindings,
        registrations: List<WhiskerElementRegistration>,
    ): Boolean {
        fun reject(message: String): Boolean {
            Log.e(LOG_TAG, "Host element bootstrap rejected: $message")
            return false
        }

        val byType = LinkedHashMap<Int, WhiskerBoundElement>()
        val byName = LinkedHashMap<String, WhiskerBoundElement>()
        for (registration in registrations) {
            val declaration = declarations[registration.name]
                ?: return reject("no Host declaration for `${registration.name}`")
            if ((registration.childPolicy == WhiskerChildPolicy.PlainText) != (declaration.factory.textUpdater != null)) {
                return reject("child policy mismatch for `${registration.name}`: Rust=${registration.childPolicy}, Host text updater=${declaration.factory.textUpdater != null}")
            }
            if (registration.textStyle != (declaration.factory.textStyleUpdater != null)) {
                return reject("text-style capability mismatch for `${registration.name}`: Rust=${registration.textStyle}, Host=${declaration.factory.textStyleUpdater != null}")
            }
            val needsHostMeasurer = registration.measurement == WhiskerMeasurement.ReplacedContent ||
                registration.measurement == WhiskerMeasurement.Custom
            if (needsHostMeasurer != (declaration.factory.measurer != null)) {
                return reject("measurement capability mismatch for `${registration.name}`: Rust=${registration.measurement}, Host measurer=${declaration.factory.measurer != null}")
            }
            val rustProps = registration.properties.associateBy { it.name }
            if (rustProps.keys != declaration.properties.keys) {
                return reject("property mismatch for `${registration.name}`: Rust=${rustProps.keys.sorted()}, Host=${declaration.properties.keys.sorted()}")
            }
            val rustEvents = registration.events.map { it.name }.toSet()
            if (rustEvents != declaration.events) {
                return reject("event mismatch for `${registration.name}`: Rust=${rustEvents.sorted()}, Host=${declaration.events.sorted()}")
            }
            val rustCommands = registration.commands.map { it.name }.toSet()
            if (rustCommands != declaration.commands.keys) {
                return reject("command mismatch for `${registration.name}`: Rust=${rustCommands.sorted()}, Host=${declaration.commands.keys.sorted()}")
            }
            val properties = registration.properties.associate { property ->
                property.id to requireNotNull(declaration.properties[property.name])
            }
            val commands = registration.commands.associate { command ->
                command.id to requireNotNull(declaration.commands[command.name])
            }
            val bound = WhiskerBoundElement(registration, declaration.factory, properties, commands)
            if (byType.put(registration.elementType, bound) != null) {
                return reject("duplicate Rust element type ${registration.elementType}")
            }
            if (byName.put(registration.name, bound) != null) {
                return reject("duplicate Rust element name `${registration.name}`")
            }
        }
        target.publish(byType)
        return true
    }
}

/** Single bootstrap boundary for independently compiled Host modules. */
public object WhiskerModuleKernel {
    private val installedNames = ConcurrentHashMap.newKeySet<String>()

    /** Validates and installs one complete service + element declaration. */
    @JvmStatic
    public fun install(module: Module, crateName: String? = null) {
        val def = module.definitionLazy
        def.validateElementDeclaration()
        val name = requireNotNull(def.name) { "ModuleDefinition requires Name" }
        val qualifiedName = module.qualifiedName
            ?: if (crateName.isNullOrEmpty() || '/' in name) name else "$crateName:$name"
        require(installedNames.add(qualifiedName)) { "module already installed: $qualifiedName" }
        module.qualifiedName = qualifiedName
        WhiskerModuleEventCenter.register(module)

        def.views.forEach { WhiskerElementRegistry.register(it, qualifiedName) }

        val functions = def.functions.associateBy { it.name }
        if (functions.isNotEmpty()) {
            WhiskerModuleRegistry.registerDispatch(qualifiedName) { method, args ->
                val function = functions[method]
                    ?: return@registerDispatch WhiskerValue.Err("unknown method `$method` on module `$name`")
                function.handler(args.asList())
            }
        }
        val asyncFunctions = def.asyncFunctions.associateBy { it.name }
        if (asyncFunctions.isNotEmpty()) {
            WhiskerModuleRegistry.registerDispatchAsync(qualifiedName) { method, args, promise ->
                val function = asyncFunctions[method] ?: return@registerDispatchAsync false
                function.handler(args.asList(), promise)
                true
            }
        }
    }
}

File: platforms/ios/Sources/WhiskerModule/WhiskerModuleRegistrar.swift
import Foundation
import UIKit

public typealias WhiskerElementEventSink = (WhiskerEventBinding, WhiskerValue) -> Void

/// Taffy available-space constraint on one measurement axis.
public enum WhiskerAvailableSpace: Equatable {
    case definite, minContent, maxContent
}

public struct WhiskerMeasureRequest {
    public let availableWidth: CGFloat?
    public let availableHeight: CGFloat?
    public let availableWidthKind: WhiskerAvailableSpace
    public let availableHeightKind: WhiskerAvailableSpace
    public let knownWidth: CGFloat?
    public let knownHeight: CGFloat?
    public let payloadVersion: UInt16
    public let payload: WhiskerValue

    public init(
        availableWidth: CGFloat?,
        availableHeight: CGFloat?,
        availableWidthKind: WhiskerAvailableSpace,
        availableHeightKind: WhiskerAvailableSpace,
        knownWidth: CGFloat?,
        knownHeight: CGFloat?,
        payloadVersion: UInt16,
        payload: WhiskerValue
    ) {
        self.availableWidth = availableWidth
        self.availableHeight = availableHeight
        self.availableWidthKind = availableWidthKind
        self.availableHeightKind = availableHeightKind
        self.knownWidth = knownWidth
        self.knownHeight = knownHeight
        self.payloadVersion = payloadVersion
        self.payload = payload
    }
}

public struct WhiskerMeasuredSize {
    public let width: CGFloat
    public let height: CGFloat

    public init(width: CGFloat, height: CGFloat) {
        self.width = width
        self.height = height
    }
}

public struct WhiskerTextContent {
    public let value: String
    public let fontFamilies: [String]
    public let fontSize: CGFloat
    public let fontWeight: Int
    public let fontStyle: WhiskerTextFontStyle
    public let lineHeight: CGFloat?
    public let letterSpacing: CGFloat
    public let fontFeatures: [WhiskerFontFeature]
    public let fontVariations: [WhiskerFontVariation]
    public let fontOpticalSizing: WhiskerFontOpticalSizing
    public let color: UIColor
    public let direction: WhiskerTextDirection
    public let alignment: WhiskerTextAlignment
    public let indent: WhiskerTextIndent
    public let wrap: Bool
    public let wordBreak: WhiskerTextWordBreak
    public let maxLines: Int
    public let overflow: WhiskerTextOverflow
    public let decoration: WhiskerTextDecoration?
    public let shadow: WhiskerTextShadow?

    public init(
        value: String,
        fontFamilies: [String] = ["system"],
        fontSize: CGFloat,
        fontWeight: Int,
        fontStyle: WhiskerTextFontStyle = .normal,
        lineHeight: CGFloat? = nil,
        letterSpacing: CGFloat = 0,
        fontFeatures: [WhiskerFontFeature] = [],
        fontVariations: [WhiskerFontVariation] = [],
        fontOpticalSizing: WhiskerFontOpticalSizing = .none,
        color: UIColor,
        direction: WhiskerTextDirection = .auto,
        alignment: WhiskerTextAlignment = .start,
        indent: WhiskerTextIndent = WhiskerTextIndent(),
        wrap: Bool = true,
        wordBreak: WhiskerTextWordBreak = .normal,
        maxLines: Int = 0,
        overflow: WhiskerTextOverflow = .clip,
        decoration: WhiskerTextDecoration? = nil,
        shadow: WhiskerTextShadow? = nil
    ) {
        self.value = value
        self.fontFamilies = fontFamilies
        self.fontSize = fontSize
        self.fontWeight = fontWeight
        self.fontStyle = fontStyle
        self.lineHeight = lineHeight
        self.letterSpacing = letterSpacing
        self.fontFeatures = fontFeatures
        self.fontVariations = fontVariations
        self.fontOpticalSizing = fontOpticalSizing
        self.color = color
        self.direction = direction
        self.alignment = alignment
        self.indent = indent
        self.wrap = wrap
        self.wordBreak = wordBreak
        self.maxLines = maxLines
        self.overflow = overflow
        self.decoration = decoration
        self.shadow = shadow
    }
}

/// Resolved inherited text style delivered independently from text content.
public struct WhiskerTextStyle {
    public let fontFamilies: [String]
    public let fontSize: CGFloat
    public let fontWeight: Int
    public let fontStyle: WhiskerTextFontStyle
    public let lineHeight: CGFloat?
    public let letterSpacing: CGFloat
    public let fontFeatures: [WhiskerFontFeature]
    public let fontVariations: [WhiskerFontVariation]
    public let fontOpticalSizing: WhiskerFontOpticalSizing
    public let color: UIColor
    public let direction: WhiskerTextDirection
    public let alignment: WhiskerTextAlignment
    public let decoration: WhiskerTextDecoration?
    public let shadow: WhiskerTextShadow?

    public init(content: WhiskerTextContent) {
        fontFamilies = content.fontFamilies
        fontSize = content.fontSize
        fontWeight = content.fontWeight
        fontStyle = content.fontStyle
        lineHeight = content.lineHeight
        letterSpacing = content.letterSpacing
        fontFeatures = content.fontFeatures
        fontVariations = content.fontVariations
        fontOpticalSizing = content.fontOpticalSizing
        color = content.color
        direction = content.direction
        alignment = content.alignment
        decoration = content.decoration
        shadow = content.shadow
    }
}

public enum WhiskerTextFontStyle: Equatable {
    case normal
    case italic
    case oblique
}

public struct WhiskerFontFeature: Equatable {
    public let tag: String
    public let value: UInt32

    public init(tag: String, value: UInt32) {
        self.tag = tag
        self.value = value
    }
}

public struct WhiskerFontVariation: Equatable {
    public let tag: String
    public let value: CGFloat

    public init(tag: String, value: CGFloat) {
        self.tag = tag
        self.value = value
    }
}

public enum WhiskerFontOpticalSizing: Equatable { case auto, none }

public enum WhiskerTextDirection: Equatable { case auto, leftToRight, rightToLeft }
public enum WhiskerTextAlignment: Equatable { case start, end, left, right, center }
public enum WhiskerTextWordBreak: Equatable { case normal, breakAll, keepAll }
public enum WhiskerTextOverflow: Equatable { case clip, ellipsis }

/** First-line indentation; percentage is relative to the final Text width. */
public struct WhiskerTextIndent: Equatable {
    public let logicalPixels: CGFloat
    public let percentage: CGFloat

    public init(logicalPixels: CGFloat = 0, percentage: CGFloat = 0) {
        self.logicalPixels = logicalPixels
        self.percentage = percentage
    }

    public func resolve(width: CGFloat) -> CGFloat {
        logicalPixels + width * percentage / 100
    }
}

public struct WhiskerTextDecoration {
    public let line: WhiskerTextDecorationLine
    public let style: WhiskerTextDecorationStyle
    public let color: UIColor

    public init(
        line: WhiskerTextDecorationLine,
        style: WhiskerTextDecorationStyle,
        color: UIColor
    ) {
        self.line = line
        self.style = style
        self.color = color
    }
}

public enum WhiskerTextDecorationLine: Equatable { case underline, lineThrough }
public enum WhiskerTextDecorationStyle: Equatable { case solid, double, dotted, dashed, wavy }

public struct WhiskerTextShadow {
    public let offset: CGSize
    public let blurRadius: CGFloat
    public let color: UIColor

    public init(offset: CGSize, blurRadius: CGFloat, color: UIColor) {
        self.offset = offset
        self.blurRadius = blurRadius
        self.color = color
    }
}

public final class WhiskerMountedElement {
    public let registration: WhiskerElementRegistration
    public let view: UIView
    private let eventSource: WhiskerEventSource?
    private let textUpdater: ((UIView, WhiskerTextContent) -> Void)?
    private let textStyleUpdater: ((UIView, WhiskerTextStyle) -> Void)?
    private let childrenHostProvider: ((UIView) -> UIView)?
    private let properties: [Int: WhiskerPropComponent]
    private let commands: [Int: WhiskerCommandComponent]
    private let eventsByName: [String: WhiskerEventBinding]
    private var eventSink: WhiskerElementEventSink
    private var eventMask: UInt64 = 0

    fileprivate init(
        registration: WhiskerElementRegistration,
        view: UIView,
        textUpdater: ((UIView, WhiskerTextContent) -> Void)?,
        textStyleUpdater: ((UIView, WhiskerTextStyle) -> Void)?,
        childrenHost: ((UIView) -> UIView)?,
        properties: [Int: WhiskerPropComponent],
        commands: [Int: WhiskerCommandComponent],
        eventsByName: [String: WhiskerEventBinding],
        eventSink: @escaping WhiskerElementEventSink
    ) {
        self.registration = registration
        self.view = view
        self.eventSource = view as? WhiskerEventSource
        self.textUpdater = textUpdater
        self.textStyleUpdater = textStyleUpdater
        self.childrenHostProvider = childrenHost
        self.properties = properties
        self.commands = commands
        self.eventsByName = eventsByName
        self.eventSink = eventSink
        installEventSink()
    }

    private func installEventSink() {
        eventSource?.installWhiskerEventSink { [weak self] name, detail in
            guard let self, let event = self.eventsByName[name] else { return }
            let bit = UInt64(1) << UInt64(event.id - 1)
            if self.eventMask & bit != 0 { self.eventSink(event, detail) }
        }
    }

    public func setProperty(_ id: Int, value: WhiskerValue) { properties[id]?.setter(view, value) }

    /// Clear is a distinct protocol operation; it is not converted to `.null`.
    public func clearProperty(_ id: Int) { properties[id]?.clearer(view) }

    public func invokeCommand(_ id: Int, parameters: WhiskerValue) {
        commands[id]?.handler(view, parameters)
    }

    public func setEventMask(_ mask: UInt64) { eventMask = mask }

    @discardableResult
    public func setText(_ content: WhiskerTextContent) -> Bool {
        guard let textUpdater else { return false }
        textUpdater(view, content)
        return true
    }

    @discardableResult
    public func setTextStyle(_ style: WhiskerTextStyle) -> Bool {
        guard let textStyleUpdater else { return false }
        textStyleUpdater(view, style)
        return true
    }

    public func childrenHost() -> UIView? { childrenHostProvider?(view) }

    /** Resets protocol-owned state before a built-in presentation is reused. */
    public func prepareForReuse(eventSink: @escaping WhiskerElementEventSink) {
        properties.values.forEach { $0.clearer(view) }
        view.isHidden = false
        eventMask = 0
        self.eventSink = eventSink
        installEventSink()
    }

    public func dispose() { eventSource?.installWhiskerEventSink(nil) }
}

/** Host-owned declaration. It contains names and behavior, never Rust IDs. */
public struct WhiskerElementFactory {
    public let name: String
    fileprivate let textUpdater: ((UIView, WhiskerTextContent) -> Void)?
    fileprivate let textStyleUpdater: ((UIView, WhiskerTextStyle) -> Void)?
    fileprivate let childrenHost: ((UIView) -> UIView)?
    fileprivate let measurer: ((WhiskerMeasureRequest) -> WhiskerMeasuredSize?)?
    fileprivate let makeView: () -> UIView

    public init(
        name: String,
        textUpdater: ((UIView, WhiskerTextContent) -> Void)? = nil,
        textStyleUpdater: ((UIView, WhiskerTextStyle) -> Void)? = nil,
        childrenHost: ((UIView) -> UIView)? = nil,
        measurer: ((WhiskerMeasureRequest) -> WhiskerMeasuredSize?)? = nil,
        makeView: @escaping () -> UIView
    ) {
        precondition(!name.isEmpty && !name.contains("@"))
        self.name = name
        self.textUpdater = textUpdater
        self.textStyleUpdater = textStyleUpdater
        self.childrenHost = childrenHost
        self.measurer = measurer
        self.makeView = makeView
    }

    fileprivate func withTextStyleUpdater(
        _ updater: ((UIView, WhiskerTextStyle) -> Void)?
    ) -> WhiskerElementFactory {
        WhiskerElementFactory(
            name: name,
            textUpdater: textUpdater,
            textStyleUpdater: updater ?? textStyleUpdater,
            childrenHost: childrenHost,
            measurer: measurer,
            makeView: makeView
        )
    }

    fileprivate func withMeasurer(
        _ provider: ((WhiskerMeasureRequest) -> WhiskerMeasuredSize?)?
    ) -> WhiskerElementFactory {
        WhiskerElementFactory(
            name: name,
            textUpdater: textUpdater,
            textStyleUpdater: textStyleUpdater,
            childrenHost: childrenHost,
            measurer: provider ?? measurer,
            makeView: makeView
        )
    }
}

private struct WhiskerDeclaredElement {
    let factory: WhiskerElementFactory
    let properties: [String: WhiskerPropComponent]
    let events: Set<String>
    let commands: [String: WhiskerCommandComponent]
}

private struct WhiskerBoundElement {
    let registration: WhiskerElementRegistration
    let factory: WhiskerElementFactory
    let properties: [Int: WhiskerPropComponent]
    let commands: [Int: WhiskerCommandComponent]
}

public enum WhiskerModuleRegistry {
    private static var modules: [String: Module] = [:]

    fileprivate static func register(_ module: Module) {
        guard let name = module.qualifiedName else { preconditionFailure("module needs a name") }
        modules[name] = module
        WhiskerModuleEventCenter.register(module)
    }

    public static func module(named name: String) -> Module? { modules[name] }
}

/** Process-wide Host declaration registry and per-surface negotiated table. */
public enum WhiskerElementRegistry {
    private static var declarations: [String: WhiskerDeclaredElement] = [:]
    private static var boundByType: [Int: WhiskerBoundElement] = [:]

    public static func register(_ factory: WhiskerElementFactory) {
        register(factory, properties: [:], events: [], commands: [:])
    }

    fileprivate static func register(_ view: WhiskerViewComponent, fallbackName: String) {
        let name = view.elementName ?? fallbackName
        if let factory = view.factory {
            let props = view.components.compactMap { $0 as? WhiskerPropComponent }
            let properties = Dictionary(uniqueKeysWithValues: props.map { ($0.name, $0) })
            let events = Set(view.components.compactMap { $0 as? WhiskerEventsComponent }.flatMap(\.names))
            let commands = Dictionary(uniqueKeysWithValues: view.components.compactMap { $0 as? WhiskerCommandComponent }.map { ($0.name, $0) })
            let textStyle = view.components.compactMap { $0 as? WhiskerTextStyleComponent }.first
            let measurement = view.components.compactMap { $0 as? WhiskerMeasurementComponent }.first
            register(
                factory.withTextStyleUpdater(textStyle?.handler).withMeasurer(measurement?.handler),
                properties: properties,
                events: events,
                commands: commands
            )
            return
        }
        guard let viewClass = view.viewClass else {
            preconditionFailure("\(name) View declaration needs a class or factory")
        }
        guard let elementType = viewClass as? WhiskerNativeElement.Type else {
            preconditionFailure("\(name) View class must implement WhiskerNativeElement")
        }
        let props = view.components.compactMap { $0 as? WhiskerPropComponent }
        let properties = Dictionary(uniqueKeysWithValues: props.map { ($0.name, $0) })
        let events = Set(view.components.compactMap { $0 as? WhiskerEventsComponent }.flatMap(\.names))
        let commands = Dictionary(uniqueKeysWithValues: view.components.compactMap { $0 as? WhiskerCommandComponent }.map { ($0.name, $0) })
        let textStyle = view.components.compactMap { $0 as? WhiskerTextStyleComponent }.first
        let measurement = view.components.compactMap { $0 as? WhiskerMeasurementComponent }.first
        register(
            WhiskerElementFactory(
                name: name,
                textStyleUpdater: textStyle?.handler,
                measurer: measurement?.handler,
                makeView: elementType.makeWhiskerView
            ),
            properties: properties,
            events: events,
            commands: commands
        )
    }

    private static func register(
        _ factory: WhiskerElementFactory,
        properties: [String: WhiskerPropComponent],
        events: Set<String>,
        commands: [String: WhiskerCommandComponent]
    ) {
        precondition(declarations[factory.name] == nil, "duplicate Host element \(factory.name)")
        declarations[factory.name] = WhiskerDeclaredElement(factory: factory, properties: properties, events: events, commands: commands)
    }

    /** Match Host strings to Rust registrations and compile compact dispatch tables. */
    @discardableResult
    public static func bind(_ registrations: [WhiskerElementRegistration]) -> Bool {
        var result: [Int: WhiskerBoundElement] = [:]
        for registration in registrations {
            guard let declaration = declarations[registration.name] else {
                return false
            }
            let needsHostMeasurer = registration.measurement == .replacedContent
                || registration.measurement == .custom
            guard registration.childPolicy.acceptsPlainText == (declaration.factory.textUpdater != nil),
                  registration.textStyle == (declaration.factory.textStyleUpdater != nil),
                  needsHostMeasurer == (declaration.factory.measurer != nil)
            else { return false }
            let rustProps = Dictionary(uniqueKeysWithValues: registration.properties.map { ($0.name, $0) })
            guard Set(rustProps.keys) == Set(declaration.properties.keys),
                  Set(registration.events.map(\.name)) == declaration.events,
                  Set(registration.commands.map(\.name)) == Set(declaration.commands.keys),
                  result[registration.elementType] == nil
            else { return false }
            let properties = Dictionary(uniqueKeysWithValues: registration.properties.map {
                ($0.id, declaration.properties[$0.name]!)
            })
            let commands = Dictionary(uniqueKeysWithValues: registration.commands.map {
                ($0.id, declaration.commands[$0.name]!)
            })
            result[registration.elementType] = WhiskerBoundElement(
                registration: registration,
                factory: declaration.factory,
                properties: properties,
                commands: commands
            )
        }
        boundByType = result
        return true
    }

    public static func mount(
        _ elementType: Int,
        eventSink: @escaping WhiskerElementEventSink
    ) -> WhiskerMountedElement? {
        guard let element = boundByType[elementType] else { return nil }
        return WhiskerMountedElement(
            registration: element.registration,
            view: element.factory.makeView(),
            textUpdater: element.factory.textUpdater,
            textStyleUpdater: element.factory.textStyleUpdater,
            childrenHost: element.factory.childrenHost,
            properties: element.properties,
            commands: element.commands,
            eventsByName: Dictionary(uniqueKeysWithValues: element.registration.events.map { ($0.name, $0) }),
            eventSink: eventSink
        )
    }

    public static func measure(
        _ elementType: Int,
        request: WhiskerMeasureRequest
    ) -> WhiskerMeasuredSize? {
        boundByType[elementType]?.factory.measurer?(request)
    }

    public static func registration(_ elementType: Int) -> WhiskerElementRegistration? {
        boundByType[elementType]?.registration
    }
}

/// Single bootstrap boundary for independently compiled Host modules.
public enum WhiskerModuleKernel {
    private static let lock = NSLock()
    private static var installedNames: Set<String> = []

    /// Validates and installs one complete service + element declaration.
    public static func install(_ module: Module) {
        let definition = module.definitionLazy
        definition.validateElementDeclaration()
        guard let name = definition.name else {
            preconditionFailure("ModuleDefinition requires Name")
        }
        if module.qualifiedName == nil { module.qualifiedName = name }
        let qualifiedName = module.qualifiedName!
        lock.lock()
        let inserted = installedNames.insert(qualifiedName).inserted
        lock.unlock()
        precondition(inserted, "module already installed: \(qualifiedName)")
        WhiskerModuleRegistry.register(module)
        definition.views.forEach { WhiskerElementRegistry.register($0, fallbackName: qualifiedName) }
    }
}
